File: LeetCode/632. Smallest Range Covering Elements from K Lists/solution.py
class Solution:
    def smallestRange(self, nums: List[List[int]]) -> List[int]:
        n = len(nums)
        mx, q = -float('inf'), []
        res = [-float('inf'), float('inf')]
        for i, arr in enumerate(nums):
            mx = max(mx, arr[0])
            q.append((arr[0],i,0))
        heapq.heapify(q)
        while q:
            val, i, j = heapq.heappop(q)
            if mx - val < res[-1] - res[0]:
                res = [val, mx]
            if j + 1 >= len(nums[i]):
                break
            mx = max(mx, nums[i][j+1])
            heapq.heappush(q, (nums[i][j+1],i,j+1))
        return res

File: LeetCode/632. Smallest Range Covering Elements from K Lists/solution .cpp
class Solution {
public:
    vector<int> smallestRange(vector<vector<int>>& nums) {
        vector<int> ans {(int)(-1e5),(int)(1e5)};
        set<tuple<int,int,int>> q;
        int k = nums.size();
        for (int i = 0; i < k; i++) q.insert({nums[i][0],i,0});
        while (q.size() == k) {
            auto [lo,i,j] = *q.begin();
            auto [hi,ii,jj] = *q.rbegin();
            if (hi-lo < ans[1]-ans[0]) ans = {lo,hi};
            q.erase(q.begin());
            if (j+1 < nums[i].size()) q.insert({nums[i][j+1],i,j+1});
        }
        return ans;
    }
};